File: src/pyodide/warp.py
"""Landmark transforms, as Coda asks for them.

The third capability on the bridge, and the cheapest one yet: it declares the same two
packages NBLAST does, so on a runtime that has already scored anything it costs a
`runPython` of these definitions and nothing else.

A thin-plate spline is two operations with very different costs, and keeping them apart is
the whole design of this file.

**Fitting** solves an `(M+4) x (M+4)` system and is cubic in the landmark count. Measured in
this runtime, single-threaded, against the shipped sets: 4,704 ms for FlyWire's 3,390
landmarks, 813 ms for MANC's 1,887, 136 ms for the hemibrain's 1,023 route to JRC2018U.
Native, the same three are 623 ms / ~110 ms / ~20 ms — so wasm costs about 7.5x here, which
is more than the 2-5x a reader would guess and worth stating rather than leaving to be
rediscovered.

**Applying** is a reduction over landmarks per point and is linear in both: 262k points per
second against 3,390 landmarks, 878k against 1,023. A 100,000-point set through the FlyWire
mirror is 382 ms.

So the fit is paid once and the result is kept — twice over. `_FITTED` holds live transforms
for the session, and `coda_warp_fit` hands the *coefficients* back so the JavaScript side can
put them in IndexedDB and return them next session through `TpsTransform.from_coefs`, which
costs 0.1-0.5 ms. Storing them as float32 costs 0.026 nm of accuracy on a value of order 1e6,
against an EM voxel of 4 nm; that was measured rather than assumed.

Both entry points take and return **float32**, which is what a `Float32Array` on the other
side is and what `SkeletonGeometry.positions` already holds. fastcore wants float64 and the
cast happens here, on the wasm heap, rather than doubling what crosses the bridge.
"""

import time

import numpy as np
import navis_fastcore as fc

#: Fitted transforms for this session, by landmark set id.
#:
#: A plain dict, never evicted. The largest entry is a 3,390-landmark spline — its source
#: points and weights are about 160 kB — and there are ten sets in total, so a session that
#: somehow touched every one of them would hold under two megabytes on a heap that is already
#: 216 MB. Evicting would mean re-paying a four-second fit to save nothing anybody would miss.
_FITTED = {}


def _as_xyz(buffer, dtype=np.float32):
    """A flat buffer from the bridge as an (N, 3) float64 array fastcore will take.

    `np.frombuffer` yields a **read-only** view, which `TpsTransform` is happy with — it reads
    its landmarks and its points and writes only into its own output. The `astype` makes the
    copy that the float64 promotion needs anyway, so nothing here is copied twice.
    """
    return np.frombuffer(buffer, dtype=dtype).reshape(-1, 3).astype(np.float64)


def _build(req):
    """The transform this request names, fitting it only if there is no other way.

    Three sources, cheapest first — the coefficients the caller brought back from IndexedDB,
    a transform already fitted in this session, and last the fit itself. The caller decides
    which of the first two it can offer; this decides nothing except the order.
    """
    key = str(req["key"])

    cached = _FITTED.get(key)
    if cached is not None:
        return cached, 0.0

    source = _as_xyz(req["source"], np.float64)

    coefs = req.get("coefficients")
    if coefs is not None:
        # `from_coefs` skips the solve entirely. No `landmarks_target`, so the result cannot
        # be negated — Coda never asks a mirror for its inverse, and a spline has no exact one
        # anyway (`__neg__` is a fresh fit in the other direction, not an inversion).
        transform = fc.TpsTransform.from_coefs(
            source,
            np.frombuffer(coefs["weights"], dtype=np.float32).reshape(-1, 3).astype(np.float64),
            np.frombuffer(coefs["affine"], dtype=np.float32).reshape(4, 3).astype(np.float64),
        )
        _FITTED[key] = transform
        return transform, 0.0

    started = time.perf_counter()
    transform = fc.TpsTransform(source, _as_xyz(req["target"], np.float64))
    _FITTED[key] = transform
    return transform, (time.perf_counter() - started) * 1000.0


def coda_warp_fit(request, report=None):
    """Fit a landmark set and hand back its coefficients, without transforming anything.

    Called when the JavaScript side has no stored coefficients for this set. It could have
    been folded into `coda_warp_apply` — and the apply path does fit when it has to — but a
    separate entry point is what lets the caller pay the four seconds *once*, write the result
    away, and then never reach this function again on any later run or in any later session.

    Returns float32, because that is what goes into the store. See the module docstring for
    what the narrowing costs.
    """
    req = request.to_py()
    if report is not None:
        report(0.05, "fitting spline (one-off)")

    transform, fit_ms = _build(req)

    if report is not None:
        report(1.0, "fitted")

    return {
        "weights": np.ascontiguousarray(transform.W, dtype=np.float32).ravel(),
        "affine": np.ascontiguousarray(transform.A, dtype=np.float32).ravel(),
        "landmarks": len(transform),
        "fitMs": fit_ms,
    }


#: Points per chunk when reporting progress.
#:
#: Chunking is **free here**, which is the opposite of the finding recorded on
#: `coda_nblast`: every point's cost is independent of every other's, so splitting the array
#: repeats no work at all. 20,000 points is about 76 ms against the largest landmark set,
#: which is a bar that moves without being a bar that costs anything.
_CHUNK = 20_000


def coda_warp_apply(request, report=None):
    """Transform points through a landmark set.

    `n_cores=1` explicitly. Pyodide has no `SharedArrayBuffer`, so `get_num_threads()` is 1
    and `None` would resolve to the same thing — passing it says which number was measured
    rather than leaving a reader to wonder whether the timings above were multi-core.
    """
    req = request.to_py()
    transform, fit_ms = _build(req)

    points = _as_xyz(req["points"])
    total = len(points)
    out = np.empty((total, 3), dtype=np.float64)

    started = time.perf_counter()
    for begin in range(0, total, _CHUNK):
        end = min(begin + _CHUNK, total)
        out[begin:end] = transform.xform(points[begin:end], 1)
        if report is not None:
            report(0.1 + 0.9 * (end / max(1, total)), f"{end:,} of {total:,} points")
    apply_ms = (time.perf_counter() - started) * 1000.0

    return {
        # float32 back, matching what the geometry buffers on the other side already are.
        "positions": np.ascontiguousarray(out, dtype=np.float32).ravel(),
        "count": total,
        "fitMs": fit_ms,
        "applyMs": apply_ms,
    }

File: src/pyodide/warp.ts
/**
 * Landmark transforms' typed wrapper over the Python bridge, and the cache that keeps the fit
 * from being paid twice.
 *
 * The third capability, written to the shape `nblast.ts` set — a `.py` in `runtime.ts`'s
 * `MODULES`, the request and result types, and functions that call `callPython` and read the
 * answer by name. What is new here is the *caching*, and it is here rather than in `warp.py`
 * because only this side can see IndexedDB.
 *
 * ## Three layers, because the fit is four seconds and the apply is a third of one
 *
 * | | costs | lives |
 * | --- | --- | --- |
 * | `_FITTED` in `warp.py` | 0 | the session |
 * | coefficients in IndexedDB | 0.1–0.5 ms to rebuild from | the browser |
 * | the fit | 136–4,704 ms | once |
 *
 * All three numbers were measured in this runtime against the shipped landmark sets. The
 * effect is that the four seconds lands exactly once per browser per landmark set — on a run
 * that is already downloading ten megabytes of CPython — and never again.
 *
 * **Landmarks stay the shipped artifact and coefficients are a derived copy**, which is the
 * important half of that arrangement. A coefficient file would be smaller and would skip the
 * fit for everybody, but it is opaque, unregenerable, and tied to fastcore's internal
 * formulation of a spline. So the store's fingerprint carries the fastcore version: a wheel
 * bump invalidates every entry rather than quietly applying last year's weights.
 */

import { cacheDelete, cacheGet, cacheSet } from '../data/cache'
import type { LandmarkPairs } from '../data/transforms/landmarks'
import { callPython } from './engine'
import type { CallOptions } from './engine'
import { float32From, numberFrom } from './types'
import sources from './sources.json'

/**
 * A fitted spline's non-affine weights and affine part, as float32.
 *
 * float32 rather than float64 because it halves what goes in the store and costs 0.026 nm
 * median / 0.061 nm maximum against the float64 fit, measured on FlyWire's 3,390 landmarks
 * over 4,000 test points. An EM voxel is 4 nm.
 */
export interface WarpCoefficients {
  /** `M * 3`, row-major. */
  weights: Float32Array
  /** `4 * 3`, row-major. Row 0 is the translation. */
  affine: Float32Array
}

/*
 * `type` rather than `interface` on anything crossing the bridge: TypeScript gives a type alias
 * an implicit index signature and an interface none, so an interface is not assignable to
 * `PyArg` and the call fails to compile with a message about `undefined`.
 */
type WarpRequest = {
  key: string
  /** `M * 3` float64, xyz interleaved, in nanometres. */
  source: Float64Array
  /** The other side of the landmark pairs. Absent when `coefficients` makes the fit unnecessary. */
  target?: Float64Array
  coefficients?: { weights: Float32Array; affine: Float32Array }
}

type ApplyRequest = WarpRequest & {
  /** `N * 3` float32, xyz interleaved. **Transferred, not copied** — see `warpPoints`. */
  points: Float32Array
}

/** What `cacheGet` compares before handing an entry back. */
function fingerprint(pairs: LandmarkPairs): string {
  // The fastcore version is in here because these are *its* coefficients: a wheel that changed
  // its kernel or its ordering would otherwise have last year's weights applied to this year's
  // spline, silently and with a perfectly plausible result.
  return `fastcore=${sources.fastcoreVersion} landmarks=${pairs.count}`
}

function cacheKey(pairs: LandmarkPairs): string {
  return `warp:${pairs.id}`
}

/**
 * Coefficients for this landmark set, from the store or by fitting.
 *
 * Never fatal on the cache: `cacheGet` resolves rather than rejecting and a miss and a broken
 * store look the same, which is what makes it safe to sit in front of a computation that can
 * always be redone. A *failed write* is likewise dropped — failing to remember something is
 * not failing to compute it.
 */
async function coefficientsFor(
  pairs: LandmarkPairs,
  options: CallOptions,
): Promise<WarpCoefficients> {
  const stored = await cacheGet<WarpCoefficients>(cacheKey(pairs), {
    fingerprint: fingerprint(pairs),
  })
  /*
   * The shape is re-checked rather than trusted. IndexedDB hands back whatever was put in it,
   * including from a build that stored a different shape under the same key — and the failure
   * downstream would be `from_coefs` raising inside Python about an array of the wrong size,
   * three layers from the cache entry that caused it.
   */
  if (
    stored?.weights instanceof Float32Array &&
    stored.affine instanceof Float32Array &&
    stored.weights.length === pairs.count * 3 &&
    stored.affine.length === 12
  ) {
    return stored
  }

  const result = await callPython(
    {
      module: 'warp',
      fn: 'coda_warp_fit',
      args: [{ key: pairs.id, source: copyOf(pairs.source), target: copyOf(pairs.target) }],
    },
    options,
  )
  const fitted: WarpCoefficients = {
    weights: float32From(result, 'weights'),
    affine: float32From(result, 'affine'),
  }
  // Awaited rather than fired and forgotten, so a run that finishes has actually written —
  // the alternative is a first mirror that pays four seconds and a second one that pays it
  // again because the write was still in flight when the tab moved on.
  await cacheSet(cacheKey(pairs), fitted, fingerprint(pairs))
  return fitted
}

/**
 * `callPython` **transfers** every typed array in a call's arguments, so anything that must
 * survive the call has to be copied first.
 *
 * The landmark buffers are memoised for the session by `landmarks.ts` and the point buffer
 * belongs to the upstream node's cached result — detaching either is a value that silently
 * becomes empty, with a viewer redrawing blank and nothing connecting that to the node that
 * ran. This is `linkage.ts`'s trap, and it bites harder here because *both* arguments are
 * borrowed rather than built for the call.
 */
function copyOf<T extends Float32Array | Float64Array>(buffer: T): T {
  return buffer.slice() as T
}

export interface WarpResult {
  /** `N * 3` float32, xyz interleaved. */
  positions: Float32Array
  /** How long the fit took, or 0 where it was already in hand. For the probe and the status bar. */
  fitMs: number
  applyMs: number
}

/**
 * Push points through a landmark set, fitting the spline once per browser.
 *
 * **`points` is consumed**: it is transferred to the worker and detached here, so a caller must
 * hand over a buffer nothing else holds. Every caller does — `warpGeometry` is the only one, and
 * it gathers into a fresh array. The landmark and coefficient buffers *are* copied, because both
 * are memoised for the session and detaching them would empty the cache they came from.
 */
export async function warpPoints(
  pairs: LandmarkPairs,
  points: Float32Array,
  options: CallOptions = {},
): Promise<WarpResult> {
  const coefficients = await coefficientsFor(pairs, options)

  // Read before the call, not after: `points` is *transferred*, so the moment `callPython`
  // posts the message this buffer is detached and its `length` reads 0. The guard below used
  // to ask the detached array how long it had been, and every warp of a non-empty geometry
  // failed with "returned N points for 0".
  const expected = points.length

  const request: ApplyRequest = {
    key: pairs.id,
    source: copyOf(pairs.source),
    coefficients: { weights: copyOf(coefficients.weights), affine: copyOf(coefficients.affine) },
    // Not copied. Every caller builds this for the call — `gatherPositions` concatenates into
    // a fresh buffer, and a multi-leg warp passes the previous leg's *result* — so the
    // transfer costs nothing and the copy was a second full-size allocation per hop: about
    // 8.7 MB for a 500-skeleton set and 36 MB for a mesh set near the ceiling.
    points,
  }
  const result = await callPython(
    { module: 'warp', fn: 'coda_warp_apply', args: [request] },
    options,
  )

  const positions = float32From(result, 'positions')
  const count = numberFrom(result, 'count')
  if (positions.length !== count * 3) {
    throw new Error(`Warp returned ${positions.length} numbers for ${count} points`)
  }
  if (count * 3 !== expected) {
    // A silent length change would scatter every coordinate after the discrepancy onto the
    // wrong point, which is a neuron that still draws.
    throw new Error(`Warp returned ${count} points for ${expected / 3}`)
  }
  return { positions, fitMs: numberFrom(result, 'fitMs'), applyMs: numberFrom(result, 'applyMs') }
}

/** Forget the stored coefficients for a landmark set. Tests; a wheel bump does it by fingerprint. */
export async function forgetWarpCoefficients(pairs: LandmarkPairs): Promise<void> {
  await cacheDelete(cacheKey(pairs))
}
